### DISCUSSION TITLE
Different way of coloring

### DISCUSSION THREAD

**Discussion by michal-barczewski** [Sakati84/3DPrintColorizer#55] [Ideas]
I've been thinking that the way coloring currently works is by pretending we have multiple materials - this causes the print to be much more complicated than it needs to be - especially small colored areas.
as in, additional walls and roofs where we change colors - which don't actually need to be there.
Instead we could simply print whole model as normal and only color specific areas.

Now - how to accomplish this:
- create model with single color - this would be printed with the 'real' extruder (optionally let 2nd extruder behave as it does now - if we want to color the base model)
- create models for each of the other colors same as before
- in cura disable 'union overlaping volumes' 'remove mesh intersection'- so generated gcode will simply overlap which normally wouldn't work - but we won't be really printing those at all
- remove supports, infill etc. for color-only models so there's only walls and roofs (stuff on which we want to draw)
- in post processing remove the 'real' extrusions that are generated and add only the coloring movements as normal

This should result should be much simpler print especially if we want complex coloring e.g. put some text on top layer.

**Comment by Sakati84** [comment]
If been trying to do this for quite some time. My idea was to slice the model 2 times. 1 time as it is done right now to get the color info and 1 time with no colors to get the more stable model. My problem is that i can´t get Cura to slice it 2 times. I doubt it is even possible from a postprocessing plugin.
So I really like your idea. To recap if I understood this correctly:
-the users imports the model as he does now
-after this he imports it a second time, assigns everthing to the first extruder and places both models in the same position
-with the right settings we get overlapping g-codes for the colors but I´m sure I could untangle them and generate the right gcode
Did I get this right?
I really like the idea but I have doubts about the additional step for the user. A lot of people probably won´t do it the right way and will blame me for it.. Any idea how this could be more fool proof / automated?

**Comment by michal-barczewski** [comment]
Hm, not quite.

[edit]
crucial part here is: you need a different set of models than you do right now.
1. a base model as you'd print with single extruder
2. a set of models for each color that overlap with the base model
[/edit]

The way i see it, you import multiple models, assign extruders, then merge - all same as you do now.
Except now (and for normal multimaterial) you have models that don't overlap, but here you'd create one base model that has all the features (a normal single material print) and multiple other models that have colored areas overlapping with the base model.
The only difference is that you have the union and mesh removal set so that cura generates gcode that overlaps.
I'd have assumed that you can still get each gcode separately same as you do now since they are assigned to different extruders.

It's possible i'm misunderstanding how the plugin works under the hood?

Also it'd make sense for this to be a separate mode - since it requires a different model than normal - this also solves (at least somewhat) the user confusion - if they don't select that mode all works as normal.

**Comment by michal-barczewski** [comment]
I've made an example model and some screenshots from cura that hopefully explain this better
[edit]
for normal print you'd use "color1" and "color2"
for new method you'd use "base" and "color2"
[/edit]

[colorizer_test_stls.zip](https://github.com/Sakati84/3DPrintColorizer/files/7048831/colorizer_test_stls.zip)
1. import models
![1-models](https://user-images.githubusercontent.com/21954392/130839472-dfe3c571-8b8d-4651-a502-65f5de1d2662.png)
2. set colors
![2-colors_set](https://user-images.githubusercontent.com/21954392/130839485-b39b65e2-1959-4bb5-84ed-0537ce8723c5.png)
3. merge models
![3-merged](https://user-images.githubusercontent.com/21954392/130839489-b31ad9a1-9569-4c65-b8a7-67288de94a05.png)
4. slice
![4-sliced](https://user-images.githubusercontent.com/21954392/130839498-e3cf135d-af82-419d-baa2-473a2192d5a2.png)

1st layer with 1st color only
![5-1st-layer-1st-color](https://user-images.githubusercontent.com/21954392/130839507-e4fb05b5-3ffa-4c14-83ed-dd299fb9fb63.png)
1st layer with both colors
![6-1st-layer-2nd-color](https://user-images.githubusercontent.com/21954392/130839515-b2a52156-7007-44c1-b0fb-2ed89f3deed3.png)

**Comment by Sakati84** [comment]
Hm.. i´m not so happy about the user needing another model. My hope was to get the base model by merging the individual models and the slice the color and the base model. But that also seems quite tricky :-(

**Comment by michal-barczewski** [comment]
Well, if one already has a multi material model all you need to do is load all parts into fusion (or w/e) and combine them into one to create the base.
If you don't already have a multi part model you have to split it anyways so it's the same work.

But i think this is a different method than normal multi color print so it should not be weird that it has it's own specific models.
Also it'd hopefully allow results that are hard or impossible to accomplish  as is - for example print a smooth top layer with ironing then draw a picture on top.

**Comment by michal-barczewski** [comment]
Honestly all this would be best implemented in cura itself as a different type of tool, and a better file format for including color information.
But that's another topic.

**Comment by michal-barczewski** [comment]
Thinking about it again, if i'm correct as to how the plugin works, which is:

for each extruder take the gcode for a layer then replace it with:

1. normal gcode with extrusion
2. tool pickup
3. duplicated normal gcode but with no extrusion, walls/top/bottom only and at an offset
4. tool return

So all that'd have to be done is an option to not include the 1st part for a specific extruder.

There is potential problem with ordering - i'm not sure cura always prints extruder 1 first for each layer.
But worst case you end up painting every other layer twice and every odd not at all - which is fine for proof of concept at least.

**Reply by Sakati84** [reply]
You are correct about your assumptions. Cura prints the extruders always in the same order

**Reply by michal-barczewski** [reply]
I just did a test with 5% (minimum allowed) flow on all extruders except 1st one and it basically worked.
Extruder moving over an already printed layer slightly distorts it but almost un-noticable.
It'd be worse on bigger areas

It's very small so kinda hard to see, but you can tell the bottom layer is continuous.
![image](https://user-images.githubusercontent.com/21954392/130848821-b14ffdb6-3d83-4a8b-9cad-c04198a335d0.png)

Another test, nevermind the crappy text, the marker is kinda worn down :)
![image](https://user-images.githubusercontent.com/21954392/130850629-3e3a14c0-3d5b-444d-af2d-c362489d410d.png)
replaced black marker for new one for last layer, which then exposes issues with marker wobbling somewhat :)
![image](https://user-images.githubusercontent.com/21954392/130850600-944f700d-15d8-4001-a852-37968fdbce39.png)

**Comment by michal-barczewski** [comment]
Couple more tests in the other orientation:

normal method 
![image](https://user-images.githubusercontent.com/21954392/130868398-bd18d50e-f72f-470f-94e5-78d25eb72dbf.png)

overlapping models @5% flow
![image](https://user-images.githubusercontent.com/21954392/130868385-a5aeea6d-66e8-4334-a88d-d94b699e2bac.png)

top down view
![image](https://user-images.githubusercontent.com/21954392/130868457-901afffe-01df-4939-87c4-eda69557b0ad.png)
![image](https://user-images.githubusercontent.com/21954392/130868767-102db688-33fa-4618-9cd1-d965ecb55b01.png)

you can still see a bit of roughness that shouldn't be there due to the 5% flow pass
it does seem to let the ink smear out a bit more - so it's possibly a trade off
the top on the other hand came out way better, i suspect due to the fact that a single pass goes over colored and non-colored area.

**Comment by Sakati84** [comment]
Ok, you convinced me. I still think most people won´t have both models or know how to use this correctly but I would like to include this as a new mode. Any ideas about how to name this in the UI? "Extrusion mode: All | Only first extruder"?

**Reply by michal-barczewski** [reply]
I think names are fine. Alternative name would be 'color-only' i guess.

As for models, there aren't that many multi-material models either tbh, and most are 2-3 color at most, so if you want to use full capability you'll need to modify them.
It might be possible to create simple combine tool that'd take a multi-material model and output the base, but even if not it's still not too hard.
I'll try to record the process.

**Reply by michal-barczewski** [reply]
https://www.youtube.com/watch?v=w973kiuiPbs

**Reply by michal-barczewski** [reply]
It might actually be useful to also add another mode (advanced?) where you can specifically select which extruders should... extrude.
This way you can for example use an existing multi-material model as is, while also adding some features which are only colored.
Or mix both approaches wherever it makes more sense.

Or instead of making separate modes simply have advanced area where you can select settings per extruder - including removing the extrusion (this could have a 'color-only mode' tooltip).

This may actually be easier to explain, as you have the following rules:
- models should either fully overlap or not at all
- disable 'union overlapping volumes' 'remove mesh intersection'
- if the sub-model does not overlap with others - use extruder with extrusion on
- if it fully overlaps - use extruder with extrusion off (color-only mode)]

You could even have two extruders for each pen - with extrusion and without.

**Comment by Anar91** [comment]
Hi everyone,
first of all thanks to Sakati84 for this cool project.
And to Michal for your contribution here. I completely agree with you and really like your idea.
I tried your method and it works great. Too bad Cura doesn't allow zero flow. Maybe the script could remove all T1 lines that have E values and only keep the offseted gcode ? I'm just not good enough at programming to try.
The only problem I encountered (and I dont know why...) is when slicing text (like your "ABC" exemple) sometime the pen doesn't lift between two letters (which draws an unwanted line). If I slice it the regular way (no overlapping meshes) it does lift between letters.
Did you notice this behavior ?
I'm still investigating but haven't found the reason so far.

Cheers.

**Reply by Anar91** [reply]
I changed "Max Comb Distance With No Retract" to 0.1 and now the pen does lift in between letters.
